### ISSUE TITLE
Ineffective checks for user initiated navigation's

### ISSUE THREAD

**Issue by NDevTK** [mjz3/LeakuidatorPlus#8]
Top level navigation's are useful as they bypass SameSite=Lax and cache partitioning while in most cases maintaining a window reference.

```javascript
async function bypassFirefox(url) {
let win = open('about:blank');
await new Promise(resolve => setTimeout(resolve, 100));
win.location.reload();
await new Promise(resolve => setTimeout(resolve, 100));
win.location = url;
return win;
}
```

```javascript
async function bypassChrome(url) {
let win = open('about:blank');
await new Promise(resolve => setTimeout(resolve, 10));
win.location = 'about:blank/#';
await new Promise(resolve => setTimeout(resolve, 10));
win.history.back();
await new Promise(resolve => setTimeout(resolve, 10));
win.location = url;
return win;
}
```

Other notes:
- Cache protection should use the header `Vary: Sec-Fetch-Dest, Sec-Fetch-Mode, Sec-Fetch-Site, Sec-Fetch-User` to avoid cache probing as recommended by https://w3c.github.io/webappsec-post-spectre-webdev/
- It has always been possible to leak what URLs someone has visited via the CSS leak https://ndev.tk/visted/ or https://www.youtube.com/watch?v=KcOQfYlyIqw
- CPU performance can leak information even without abusing the CPU cache https://devicemonitor.glitch.me/ this is a leak I wanted to add to the XSLeaks wiki but never finished such attacks probably should be added.
- "Math.random() does not provide cryptographically secure random numbers. Do not use them for anything related to security. "

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-Fetch-User

**Comment by mjz3** [comment]
Thank you for raising these issues NDevTK.
As far as I understood, the examples provided above should be related to popunder and tabunder variants of the attacks we introduced in the paper.
Can you explain how these two examples (for Firefox and Chrome) would bypass the protections provided by Leakuidator+  for popunder and tabunder variants? (Specially, where it records the relations between different tabs and windows and applies the protection on related tabs and windows.)

**Comment by NDevTK** [comment]
Tab relations are cut to soon meaning no protection is provided.
history.back() is of type forward_back on chrome.
https://github.com/mjz3/LeakuidatorPlus/blob/afb01d73f7d8f5cb6a935839c72cf1c50d12cc24/ext/background.js#L205

**Comment by mjz3** [comment]
The webNavigation Transition qualifier documentation for "forward_back"	 says "The **user** used the Forward or Back button to initiate the navigation."
We exclude "forward_back" because it is **user** initiated. It is interesting that JavaScript initiated history.back() function escapes this!
I will test this to verify.

**Comment by mjz3** [comment]
I implemented your suggested approach in Chrome using the popunder method from [1], trying to load an authenticated Twitter page. The attack is protected as expected.

Also, as I explained in the previous comment, this bypass shouldn't work since it is initiated by JavaScript.

You can try it yourself by modifying the attack page in [1]. If you are able to bypass the protection, please share with me the modified page so I can verify as well.

[1] https://github.com/leakuidatorplusteam/artifacts/blob/main/wp_popunder.html

**Comment by NDevTK** [comment]
It does not matter if it shouldn't work it does.
https://github.com/NDevTK/artifacts/blob/main/wp_popunder.html

**Comment by mjz3** [comment]
I tested this attack page with Chrome. It loads the leaky resource in a new tab in foreground and bypasses the protection. It is valid, though not stealthy.

You are welcome to make any suggestions to protect against this bypass.

**Comment by NDevTK** [comment]
Yeah its not going to be stealth without a popunder which there are some, This issue is not related to the bypass directly.
